sigstore-java/src/main/java/dev/sigstore/tuf/Updater.java

```diff
@@ -225,82 +234,190 @@ void verifyDelegate(
+    }
+    var targetsResult = targetsResultMaybe.get();
     // 2) check hash against snapshot.targets.hash, else fail.
-
+    verifyHashes(
+        "targets.json",
+        targetsResult.getRawBytes(),
+        snapshot.getSignedMeta().getTargetMeta("targets.json").getHashes());
     // 3) check against threshold of keys as specified by trusted root.json
-
+    verifyDelegate(root, targetsResult.getMetaResource());
     // 4) check targets.version == snapshot.targets.version, else fail.
-
+    int targetsVersion = targetsResult.getMetaResource().getSignedMeta().getVersion();
+    int snapshotTargetsVersion =
+        snapshot.getSignedMeta().getTargetMeta("targets.json").getVersion();
+    if (targetsVersion != snapshotTargetsVersion) {
+      throw new SnapshotVersionMismatchException(snapshotTargetsVersion, targetsVersion);
+    }
     // 5) check expiration is after tuf update start time
-
+    throwIfExpired(targetsResult.getMetaResource().getSignedMeta().getExpiresAsDate());
     // 6) persist targets metadata
+    // why do we persist the
+    localStore.storeMeta(targetsResult.getMetaResource());
+    return targetsResult.getMetaResource();
+  }
 
-    // 7) starting at each top level target role:
-    //        do pre-order DFS of metadata.
-    //        skip already visited roles
-    //        if maximum roles visited go to 8) downloading targets
-    //        process delegations (not sure if we need this yet)
-
-    // 8) If target is missing metadata fail.
-
-    // 9) Download target up to length specified in bytes. verify against hash.
-
-    // Done!!
-    return null;
+  void downloadTargets(Targets targets)
+      throws IOException, TargetMetadataMissingException, FileNotFoundException {
+    // Skip #7 and go straight to downloading targets. It looks like delegations were removed from
+    // sigstore TUF data.
+    // {@see https://github.com/sigstore/sigstore/issues/562}
+    // TODO(patrick@chainguard.dev): get review from Asra on whether we should be processing
+    // delegates.
+    for (Map.Entry<String, TargetMeta.TargetData> entry :
+        targets.getSignedMeta().getTargets().entrySet()) {
+      String targetName = entry.getKey();
+      // 8) If target is missing metadata fail.
+      // Note: This can't actually happen due to the way GSON is setup the targets.json would fail
+      // to parse. Leaving
+      // this code in in-case we eventually allow it in de-serialization.
+      if (entry.getValue() == null) {
+        throw new TargetMetadataMissingException(targetName);
+      }
+      TargetMeta.TargetData targetData = entry.getValue();
+      // 9) Download target up to length specified in bytes. verify against hash.
+      // TODO(patrick@chainguard.dev): Update this code to use consistent snapshots.
+      // e.g. "{targetData.getHashes().getSha512()}.{targetName}"
+      var targetBytes = fetcher.fetchResource("targets/" + targetName, targetData.getLength());
```

Comment:
Should targetName be url-encoded?
Url concatenation sounds unsafe

Reply:
:thinking: the fetchResource implementation passes the path to GenericUrl which will also concatenates the path simply as a string to the mirror context.  I presumed that GenericUrl handles the url-encoding and looking into the code it appears to do so (despite the parameter name).  The input is trusted, what are your exact thoughts on the dangers and what do you have in mind as an alternative?  Building every component of the path from URLs?

Reply:
I might need looking closer, however, building urls via string concatenation is always suspicious